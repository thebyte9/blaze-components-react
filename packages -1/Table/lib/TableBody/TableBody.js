"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var checkboxes_1 = __importDefault(require("@blaze-react/checkboxes"));
var utils_1 = __importDefault(require("@blaze-react/utils"));
var react_1 = __importDefault(require("react"));
var TableBody = function (_a) {
    var allRows = _a.allRows, checkboxes = _a.checkboxes, identification = _a.identification, selected = _a.selected, handleSelected = _a.handleSelected, columns = _a.columns, placeholder = _a.placeholder, uniqueId = _a.utils.uniqueId;
    return (react_1["default"].createElement("tbody", null,
        allRows.map(function (row, key) { return (react_1["default"].createElement("tr", { key: uniqueId(row), "data-testid": "tablerow-" + (key + 1) },
            checkboxes && (react_1["default"].createElement("td", null,
                react_1["default"].createElement(checkboxes_1["default"], { "data-testid": "row-checkbox-" + (key + 1), options: {
                        checked: selected.includes(row[identification]),
                        id: row[identification],
                        value: row[identification]
                    }, onChange: function (_a) {
                        var value = _a.value;
                        return handleSelected(value, row[identification]);
                    } }))),
            columns.map(function (column) { return (react_1["default"].createElement("td", { key: column }, row[column])); }))); }),
        !allRows.length && (react_1["default"].createElement("tr", null,
            react_1["default"].createElement("td", { colSpan: checkboxes ? columns.length + 1 : columns.length, align: "center" }, placeholder)))));
};
exports["default"] = utils_1["default"](TableBody);
