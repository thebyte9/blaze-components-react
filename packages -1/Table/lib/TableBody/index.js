"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var TableBody_1 = __importDefault(require("./TableBody"));
exports["default"] = TableBody_1["default"];
