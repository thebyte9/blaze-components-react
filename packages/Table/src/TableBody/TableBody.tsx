import { Checkboxes } from "@blaze-react/checkboxes";
import withUtils from "@blaze-react/utils";
import React from "react";
import VirtualList from "react-tiny-virtual-list";
interface ICheckbox {
  checked: boolean;
  id: string | number;
  value: any;
}
interface ITableBody {
  allRows: object[];
  checkboxes?: boolean;
  identification: string;
  selected: any[];
  handleSelected: (
    checked: ICheckbox[],
    value: string | ICheckbox[],
    isMultiselect?: boolean
  ) => void;
  columns: string[];
  placeholder: string | JSX.Element;
  utils: {
    uniqueId: (element: any) => string;
  };
  bodyRef: any;
  scrollToIndex: number;
  overScanBuffer: number;
  onRenderItems?: (arg: any) => void;
  onClickRow: (arg: any) => void;
}

const TableBody = ({
  allRows,
  checkboxes,
  identification,
  selected,
  handleSelected,
  columns,
  placeholder,
  utils: { uniqueId },
  bodyRef,
  overScanBuffer = 20,
  onRenderItems,
  scrollToIndex = 0,
  onClickRow
}: ITableBody): JSX.Element => {
  return (
    <div ref={bodyRef} className="table-body">
      {bodyRef.current && allRows.length && (
        <VirtualList
          width={"100%"}
          height={bodyRef.current.offsetHeight}
          itemCount={allRows.length}
          itemSize={62}
          scrollToIndex={scrollToIndex}
          overscanCount={overScanBuffer}
          onItemsRendered={onRenderItems}
          renderItem={({ index, style }) => (
            <div
              onClick={() => onClickRow({ ...allRows[index], index })}
              className="table-row"
              key={uniqueId(allRows[index])}
              data-testid={`tablerow-${index + 1}`}
              style={style}
            >
              {checkboxes && (
                <div className="table-cell--checkbox">
                  <Checkboxes
                    test-id={`row-checkbox-${index + 1}`}
                    options={{
                      checked: selected.includes(
                        allRows[index][identification]
                      ),
                      id: allRows[index][identification],
                      value: allRows[index][identification]
                    }}
                    onChange={({ value }: { value: ICheckbox[] }): void =>
                      handleSelected(value, allRows[index][identification])
                    }
                  />
                </div>
              )}
              {columns.map(
                (column: string): JSX.Element => (
                  <div className="table-cell" key={column}>
                    <div className="table-cell--content">
                      {allRows[index][column]}
                    </div>
                  </div>
                )
              )}
            </div>
          )}
        />
      )}

      {!allRows.length && (
        <div className="table-body-placeholder">
          <div>{placeholder}</div>
        </div>
      )}
    </div>
  );
};

export default withUtils(TableBody);
