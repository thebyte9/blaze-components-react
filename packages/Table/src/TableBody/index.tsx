import Checkboxes from "@blaze-react/checkboxes";
import utils from "@blaze-react/utils";
import React from "react";

interface ITableBody {
  allRows: any;
  checkboxes: any;
  identification: any;
  selected: any;
  handleSelected: any;
  columns: any;
  placeholder: any;
  colSpan: any;
  utils: { uniqueId: any };
}

const TableBody = ({
  allRows,
  checkboxes,
  identification,
  selected,
  handleSelected,
  columns,
  placeholder,
  colSpan,
  utils: { uniqueId }
}: ITableBody) => (
  <tbody>
    {allRows.map((row: any) => (
      <tr key={uniqueId(row)}>
        {checkboxes && (
          <td>
            <Checkboxes
              withEffect
              options={[
                {
                  checked: selected.includes(row[identification]),
                  id: row[identification],
                  value: row[identification]
                }
              ]}
              onChange={({ checked }: any) =>
                handleSelected(checked, row[identification])
              }
            />
          </td>
        )}
        {columns.map((column: any) => (
          <td key={column}>{row[column]}</td>
        ))}
      </tr>
    ))}
    {!allRows.length && (
      <tr>
        <td colSpan={colSpan} align="center">
          {placeholder}
        </td>
      </tr>
    )}
  </tbody>
);

export default utils(TableBody);
