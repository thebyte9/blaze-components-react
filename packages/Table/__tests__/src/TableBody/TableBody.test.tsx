import { cleanup, fireEvent, render } from "@testing-library/react";
import expect from "expect";
import React from "react";
import TableBody from "../../../src/TableBody/TableBody";
import { data } from "../../mocks";

describe("Table body", () => {
  afterEach(cleanup);

  it("should be defined", () => {
    expect(TableBody).toBeDefined();
  });

  it("should show placeholder if there is no data yet", () => {
    const placeholder = "The table is empty of records";

    const { getByText } = render(
      <TableBody
        allRows={[]}
        bodyRef={{ current: false }}
        placeholder={placeholder}
      />
    );

    getByText(placeholder);
  });

  it("should render data and pass it throught virtual list", () => {
    const { container } = render(
      <TableBody
        selected={[]}
        onItemsRendered={jest.fn}
        checkboxes={true}
        columns={data.columns}
        allRows={data.rows}
        bodyRef={{ current: { offsetHeight: 1000 } }}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should retun row on be clicked", () => {
    const mockedOnClickRow = jest.fn();
    const { getByTestId } = render(
      <TableBody
        handleSelected={jest.fn()}
        onClickRow={mockedOnClickRow}
        selected={[]}
        onItemsRendered={jest.fn}
        checkboxes={true}
        columns={data.columns}
        allRows={data.rows}
        bodyRef={{ current: { offsetHeight: 1000 } }}
      />
    );

    const row = getByTestId("row-checkbox-2");
    fireEvent.click(row);

    expect(mockedOnClickRow).toHaveBeenCalledWith({
      ...data.rows[1],
      index: 1
    });
  });
});
