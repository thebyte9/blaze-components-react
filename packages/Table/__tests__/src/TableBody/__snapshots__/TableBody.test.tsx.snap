// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table body should render data and pass it throught virtual list 1`] = `
<div>
  <div
    class="table-body"
  >
    <div
      style="overflow: auto; will-change: transform; height: 1000px; width: 100%;"
    >
      <div
        style="position: relative; width: 100%; min-height: 100%; height: 124px;"
      >
        <div
          class="table-row"
          data-testid="tablerow-1"
          style="position: absolute; top: 0px; left: 0px; width: 100%; height: 62px;"
        >
          <div
            class="table-cell--checkbox"
          >
            <div
              class="form-field form-field--checkbox"
              data-testid="checkbox-1"
              role="button"
            >
              <input
                class="form-checkbox"
                data-testid="row-checkbox-1"
                id="uid-eyJjaGVja2VkIjpmYWxzZX0="
                readonly=""
                type="checkbox"
                value=""
              />
              <label
                for="uid-eyJjaGVja2VkIjpmYWxzZX0="
              />
            </div>
          </div>
          <div
            class="table-cell"
          >
            <div
              class="table-cell--content"
            >
              1
            </div>
          </div>
          <div
            class="table-cell"
          >
            <div
              class="table-cell--content"
            >
              Lorem
            </div>
          </div>
          <div
            class="table-cell"
          >
            <div
              class="table-cell--content"
            >
              52
            </div>
          </div>
        </div>
        <div
          class="table-row"
          data-testid="tablerow-2"
          style="position: absolute; top: 62px; left: 0px; width: 100%; height: 62px;"
        >
          <div
            class="table-cell--checkbox"
          >
            <div
              class="form-field form-field--checkbox"
              data-testid="checkbox-1"
              role="button"
            >
              <input
                class="form-checkbox"
                data-testid="row-checkbox-2"
                id="uid-eyJjaGVja2VkIjpmYWxzZX0="
                readonly=""
                type="checkbox"
                value=""
              />
              <label
                for="uid-eyJjaGVja2VkIjpmYWxzZX0="
              />
            </div>
          </div>
          <div
            class="table-cell"
          >
            <div
              class="table-cell--content"
            >
              2
            </div>
          </div>
          <div
            class="table-cell"
          >
            <div
              class="table-cell--content"
            >
              Ipsum
            </div>
          </div>
          <div
            class="table-cell"
          >
            <div
              class="table-cell--content"
            >
              43
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
